File: packages/backstage-ts/src/script-registry.ts
/**
 * Backstage SDK - Script Registry
 *
 * Manages Lua scripts for Redis execution with SHA caching and automatic
 * reload on NOSCRIPT errors.
 */

import { type RedisClient } from './types';

export interface ScriptDef {
  /** Lua script source code */
  script: string;
  /** Map of key names to their 1-based KEY index */
  keys: Record<string, number>;
}

/**
 * Manages Lua scripts for Redis execution.
 * Handles loading scripts and executing them via EVALSHA for performance.
 */
export class ScriptRegistry {
  private redis: RedisClient;
  private scripts: Map<string, { sha: string; def: ScriptDef }> = new Map();

  constructor(redis: RedisClient) {
    this.redis = redis;
  }

  /**
   * Load multiple Lua scripts into Redis and the registry.
   *
   * Scripts are loaded using Redis SCRIPT LOAD command which returns
   * a SHA digest. This SHA is used for subsequent EVALSHA calls.
   *
   * @param scripts Map of script names to script definitions
   * @example
   * ```typescript
   * await registry.load({
   *   claimRide: {
   *     script: `
   *       local status = redis.call('HGET', KEYS[1], 'status')
   *       if status ~= 'pending' then return {err='already_claimed'} end
   *       redis.call('HSET', KEYS[1], 'status', 'accepted', 'driverId', ARGV[1])
   *       return {ok=ARGV[2]}
   *     `,
   *     keys: { rideStateKey: 1 }
   *   }
   * });
   * ```
   */
  async load(scripts: Record<string, ScriptDef>): Promise<void> {
    for (const [name, def] of Object.entries(scripts)) {
      const sha = (await this.redis.send('SCRIPT', [
        'LOAD',
        def.script,
      ])) as string;
      this.scripts.set(name, { sha, def });
    }
  }

  /**
   * Run a registered Lua script.
   *
   * Keys are mapped by name - the order is determined by the indices
   * specified in the script definition.
   *
   * @param name Name of the script to run (must be registered via load)
   * @param keys Map of key names to their Redis key values
   * @param args Arguments to pass to the script (ARGV[1], ARGV[2], etc.)
   * @returns Script execution result (parsed from Redis response)
   *
   * @example
   * ```typescript
   * const result = await registry.run<{ ok?: string; err?: string }>(
   *   'claimRide',
   *   { rideStateKey: `ride:state:${rideId}` },
   *   [driverId, eventId]
   * );
   *
   * if (result.ok) {
   *   console.log('Ride claimed successfully');
   * } else if (result.err) {
   *   console.error('Failed:', result.err);
   * }
   * ```
   */
  async run<T = unknown>(
    name: string,
    keys: Record<string, string>,
    args: (string | number)[],
  ): Promise<T> {
    const entry = this.scripts.get(name);
    if (!entry) {
      throw new Error(`Script '${name}' is not registered. Call load() first.`);
    }

    const { sha, def } = entry;

    // Build ordered key array based on definition indices
    const keyCount = Object.keys(def.keys).length;
    const orderedKeys = new Array<string>(keyCount);

    for (const [keyName, index] of Object.entries(def.keys)) {
      const value = keys[keyName];
      if (value === undefined) {
        throw new Error(
          `Missing required key '${keyName}' for script '${name}'. ` +
            `Expected keys: ${Object.keys(def.keys).join(', ')}`,
        );
      }
      // Indices are 1-based, array is 0-based
      orderedKeys[index - 1] = value;
    }

    // Validate no gaps in key indices
    for (let i = 0; i < orderedKeys.length; i++) {
      if (orderedKeys[i] === undefined) {
        throw new Error(
          `Invalid key definition for script '${name}'. ` +
            `Missing key for index ${i + 1}. ` +
            `Ensure all indices from 1 to ${keyCount} are defined.`,
        );
      }
    }

    const commandArgs = [
      sha,
      String(orderedKeys.length),
      ...orderedKeys,
      ...args.map(String),
    ];

    try {
      return (await this.redis.send('EVALSHA', commandArgs)) as T;
    } catch (err: unknown) {
      // Handle NOSCRIPT error (Redis was restarted, lost cached scripts)
      if (err instanceof Error && err.message.includes('NOSCRIPT')) {
        // Reload and retry once
        const newSha = (await this.redis.send('SCRIPT', [
          'LOAD',
          def.script,
        ])) as string;
        this.scripts.set(name, { sha: newSha, def });
        commandArgs[0] = newSha;
        return (await this.redis.send('EVALSHA', commandArgs)) as T;
      }
      throw err;
    }
  }

  /**
   * Get information about a registered script.
   *
   * @param name Script name
   * @returns Script definition or undefined if not found
   */
  get(name: string): { sha: string; def: ScriptDef } | undefined {
    return this.scripts.get(name);
  }

  /**
   * Check if a script is registered.
   *
   * @param name Script name
   * @returns True if script is registered
   */
  has(name: string): boolean {
    return this.scripts.has(name);
  }

  /**
   * Get all registered script names.
   *
   * @returns Array of script names
   */
  keys(): string[] {
    return Array.from(this.scripts.keys());
  }
}

File: packages/backstage-go/script_registry.go
// Package backstage script registry.
// Manages loading and executing Lua scripts in Redis with automatic SHA caching and error recovery.
package backstage

import (
	"context"
	"fmt"
	"strings"

	"github.com/redis/go-redis/v9"
)

// ScriptDef defines a Lua script and its expected keys.
type ScriptDef struct {
	// Script is the Lua script source code.
	Script string
	// Keys maps key names to their 1-based KEY index in the Redis call.
	// For example, {"myKey": 1} means KEYS[1] will correspond to the value provided for "myKey".
	Keys map[string]int
}

// registeredScript holds the internal state of a loaded script.
type registeredScript struct {
	sha string
	def ScriptDef
}

// ScriptRegistry manages Lua scripts for Redis execution using EVALSHA.
type ScriptRegistry struct {
	client  redis.UniversalClient
	scripts map[string]*registeredScript
}

// NewScriptRegistry creates a new ScriptRegistry.
func NewScriptRegistry(client redis.UniversalClient) *ScriptRegistry {
	return &ScriptRegistry{
		client:  client,
		scripts: make(map[string]*registeredScript),
	}
}

// Load loads multiple scripts into Redis and registers them.
func (r *ScriptRegistry) Load(ctx context.Context, scripts map[string]ScriptDef) error {
	for name, def := range scripts {
		sha, err := r.client.ScriptLoad(ctx, def.Script).Result()
		if err != nil {
			return fmt.Errorf("failed to load script %q: %w", name, err)
		}
		r.scripts[name] = &registeredScript{sha: sha, def: def}
	}
	return nil
}

// Run executes a registered script.
// keys is a map of key names (as defined in ScriptDef.Keys) to their actual Redis key values.
// args is a list of arguments to pass to the script (ARGV[1], ARGV[2], ...).
func (r *ScriptRegistry) Run(ctx context.Context, name string, keys map[string]string, args ...interface{}) (interface{}, error) {
	script, ok := r.scripts[name]
	if !ok {
		return nil, fmt.Errorf("script %q is not registered", name)
	}

	// Prepare KEYS array
	numKeys := len(script.def.Keys)
	orderedKeys := make([]string, numKeys)

	for keyName, index := range script.def.Keys {
		val, ok := keys[keyName]
		if !ok {
			expectedKeys := make([]string, 0, len(script.def.Keys))
			for k := range script.def.Keys {
				expectedKeys = append(expectedKeys, k)
			}
			return nil, fmt.Errorf("missing required key %q for script %q. Expected keys: %s", keyName, name, strings.Join(expectedKeys, ", "))
		}
		
		// Validate index range
		if index < 1 || index > numKeys {
			return nil, fmt.Errorf("invalid key index %d for key %q in script %q. Indices must be between 1 and %d", index, keyName, name, numKeys)
		}

		orderedKeys[index-1] = val
	}

	// Verify no gaps
	for i, k := range orderedKeys {
		if k == "" {
			return nil, fmt.Errorf("missing key for index %d in script %q", i+1, name)
		}
	}

	// Execute
	res, err := r.client.EvalSha(ctx, script.sha, orderedKeys, args...).Result()
	if err != nil {
		// Handle NOSCRIPT error
		if strings.HasPrefix(err.Error(), "NOSCRIPT") {
			// Reload
			newSha, loadErr := r.client.ScriptLoad(ctx, script.def.Script).Result()
			if loadErr != nil {
				return nil, fmt.Errorf("failed to reload script %q after NOSCRIPT error: %w", name, loadErr)
			}
			script.sha = newSha
			// Retry
			return r.client.EvalSha(ctx, newSha, orderedKeys, args...).Result()
		}
		return nil, err
	}

	return res, nil
}

// Has checks if a script is registered.
func (r *ScriptRegistry) Has(name string) bool {
	_, ok := r.scripts[name]
	return ok
}

// GetSHA returns the SHA of a registered script, or empty string if not found.
func (r *ScriptRegistry) GetSHA(name string) string {
	if s, ok := r.scripts[name]; ok {
		return s.sha
	}
	return ""
}
